aws/middleware/endpoints.go

```diff
@@ -0,0 +1,69 @@
+package middleware
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/awslabs/smithy-go/middleware"
+	smithyhttp "github.com/awslabs/smithy-go/transport/http"
+)
+
+// ResolveServiceEndpoint is a middleware that will resolve the service endpoint.
+type ResolveServiceEndpoint struct {
+	Resolver aws.EndpointResolver
+}
+
+// ID is a middleware identifier
+func (r ResolveServiceEndpoint) ID() string {
+	return "ResolveServiceEndpoint"
+}
+
+// HandleSerialize resolves the service endpoint and returns
+func (r ResolveServiceEndpoint) HandleSerialize(
+	ctx context.Context, in middleware.SerializeInput, next middleware.SerializeHandler,
+) (out middleware.SerializeOutput, metadata middleware.Metadata, err error) {
+	req, ok := in.Request.(*smithyhttp.Request)
+	if !ok {
+		return out, metadata, fmt.Errorf("unknown transport type %T", in.Request)
+	}
+
+	var endpoint aws.Endpoint
+	endpoint, err = r.Resolver.ResolveEndpoint(GetEndpointID(ctx), GetRegion(ctx))
```

Comment:
nit probably want to name `endpointId` endpointPrefix for now. Hopefully we can replace this with service id in the future.